pdns/recursordist/lwres.cc

```diff
@@ -287,22 +321,41 @@ static void logIncomingResponse(const std::shared_ptr<std::vector<std::unique_pt
   }
 }
 
-static bool tcpconnect(const ComboAddress& ip, TCPOutConnectionManager::Connection& connection, bool& dnsOverTLS, const std::string& nsName)
+class BindError
 {
-  dnsOverTLS = SyncRes::s_dot_to_port_853 && ip.getPort() == 853;
+};
 
-  connection = t_tcp_manager.get(ip);
+static bool tcpconnect(const OptLog& log, const ComboAddress& remote, const std::optional<ComboAddress> localBind, TCPOutConnectionManager::Connection& connection, bool& dnsOverTLS, const std::string& nsName)
+{
+  dnsOverTLS = SyncRes::s_dot_to_port_853 && remote.getPort() == 853;
+
+  connection = t_tcp_manager.get(std::make_pair(remote, localBind));
   if (connection.d_handler) {
     return false;
   }
 
   const struct timeval timeout{
     g_networkTimeoutMsec / 1000, static_cast<suseconds_t>(g_networkTimeoutMsec) % 1000 * 1000};
-  Socket s(ip.sin4.sin_family, SOCK_STREAM);
-  s.setNonBlocking();
-  setTCPNoDelay(s.getHandle());
-  ComboAddress localip = pdns::getQueryLocalAddress(ip.sin4.sin_family, 0);
-  s.bind(localip);
+  Socket sock(remote.sin4.sin_family, SOCK_STREAM);
+  sock.setNonBlocking();
+  setTCPNoDelay(sock.getHandle());
+  ComboAddress localip = localBind ? *localBind : pdns::getQueryLocalAddress(remote.sin4.sin_family, 0);
+  if (localBind) {
```

Comment:
right, because reusing a cookie provides more protection than switching IP, makes sense

Reply:
I'll add a comment referring to the RFC that explains this